src/typecheck/destructuring.rs

```diff
@@ -1,56 +1,47 @@
 use std::collections::HashMap;
 
 use crate::{
-    destruct::{Destruct, Match},
+    destructuring::{DestructuredRecord, Match},
     identifier::Ident,
     types::{RecordRowF, RecordRowsF, TypeF},
 };
 
-use super::{
-    mk_uniftype, Environment, GenericUnifRecordRowsIteratorItem, GenericUnifType, State,
-    UnifRecordRows, UnifType,
-};
+use super::{Environment, GenericUnifRecordRowsIteratorItem, State, UnifRecordRows, UnifType};
 
 /// Build a `UnifType` from a `Destruct` pattern, creating fresh unification
 /// variables wherever specific types are unknown.
-pub fn build_pattern_type(state: &mut State, pat: &Destruct) -> UnifType {
-    match pat {
-        Destruct::Record { matches, open, .. } => {
-            let tail = if *open {
-                state.table.fresh_rrows_uvar()
-            } else {
-                UnifRecordRows::Concrete(RecordRowsF::Empty)
-            };
-
-            let rows = matches.iter().map(|m| match m {
-                Match::Simple(id, _) => RecordRowF {
-                    id: *id,
-                    types: Box::new(state.table.fresh_type_uvar()),
-                },
-                Match::Assign(id, _, (_, Destruct::Empty)) => RecordRowF {
-                    id: *id,
-                    types: Box::new(state.table.fresh_type_uvar()),
-                },
-                Match::Assign(id, _, (_, r_pat @ Destruct::Record { .. })) => {
-                    let ty = build_pattern_type(state, r_pat);
-                    RecordRowF {
-                        id: *id,
-                        types: Box::new(ty),
-                    }
-                }
-            });
-
-            let rrows = rows.fold(tail, |tail, row| {
-                UnifRecordRows::Concrete(RecordRowsF::Extend {
-                    row,
-                    tail: Box::new(tail),
-                })
-            });
-
-            UnifType::Concrete(TypeF::Record(rrows))
+pub fn build_pattern_type(state: &mut State, pat: &DestructuredRecord) -> UnifRecordRows {
```

Comment:
I think `build_pattern_type` has to act differently, depending on if we are in strict mode or not. Otherwise, your pattern will infer "too much" type in walking mode:

```
fun {x} => (x : Num)
```

Right now this typechecks, while it shouldn't (for example `fun x => (x : Num)` doesn't). `x` is introduced outside of any typed block, so it should be given type `Dyn`. Note that this also happens with intermediate pattern:

```
fun {x = foo @ {bar | Num}} => ...
```

So, `build_pattern_type` should probably does as `binding_type`: depending on being in strict mode or not, it should affect either `Dyn` or a unification variable to the leaf of the pattern.

Reply:
Ah yeah, good point. In the first version this worked because we used the `Dyn` variant. I'll have a look at how `binding_type` does this & try to copy the pattern used there.

Reply:
It might be dead simple: take an additional `strict: bool` argument (well, ideally, `mode: TypecheckMode`, but this is not the only place where this change needs to be made, so we can do that in a pass in a later PR). If strict, when reaching the "leaf" of a pattern, produces a fresh uvar, as it is now. In non-strict mode, assign `Dyn` instead.

Reply:
Yeah, my only concern there is that passing a flag like that is a form of [control coupling](https://thoughtbot.com/blog/types-of-coupling#control-coupling), which is maybe fine, but equally we could just provide another function like `walk_pattern_type` & let the caller decide which is most appropriate.

Reply:
I guess you still want to have only one implementation though, isn't it? Otherwise we'll have to duplicate code, which is bad  (I didn't know about control coupling! That being said, it seems to be about OOP, where "the object should decide what to do based on their own state", but at first sight it doesn't seem to fully translate to the functional style).

In the end I find `f("hello")/f_foo("hello", Mode::Foo)` and `f("hello")/f_bar("hello", Mode::Bar)` mostly isomorphic, but if your point is that we should wrap the one with the flag in two different functions for strict and non strict mode, I'm totally fine with that! That's probably cleaner, yes. Just have to find a good name, which is the easy part :roll_eyes: 

As a side note, I'm afraid `walk_pattern_type`  is a bit misleading, as we don't actually walk the pattern. It's probably more something like `build_pattern_in_walk_mode` and `build_pattern_in_check_mode` or in strict/nonstrict mode. The latter distinction might sound a bit odd at first, but I think it might be more precise, because with bidirectional typing, it would be fair to say we actually have 3 modes: walk, check, and infer. So "checking" is ambiguous. Maybe walk/typechecking is also fine, instead of nonstrict/strict.

Reply:
Yeah in my mind we'd maintain a single implementation, but it'd be internal to the `typecheck::destructuring` module, and we'd expose two functions, (with significantly better names than) `build_pattern_with_unif_var_leaves` and `build_pattern_with_dyn_leaves`.

Another possibility which is _maybe_ a little less coupled is to take a `() -> UnifType` argument which decides what to do at each pattern leaf. That way the `destructuring` module owns the algorithm, but the caller can control the behaviour as required. 

I'll play around with this & hopefully push something later today, though I'm a little busy with client work right now so can't make any promises.

Reply:
As a datapoint, I think the `() -> UnifType` sounds a bit overkill. In practice it seems we only have to handle two behaviors, so in the end we will just end up having to make a closure inside typechecking. On the other hand just calling a specialized function makes the intent clearer IMHO.

For the private function though, both approach (boolean, enum, or closure) are reasonable, I guess.